src/executionTimingsReporter.ts

```diff
@@ -0,0 +1,139 @@
+import { HighResTimer, calculateElapsedTime } from "./utils";
+import { EventEmitter } from "events";
+
+export type TimingsReport = {[stepName: string]: [number] | number};
+
+export const stepStartedEventName = 'stepStarted';
+export const milestoneReachedEventName = 'milestoneReached';
+
+export interface StepStartedEventArguments {
+    stepName: string;
+}
+
+export interface MilestoneReachedEventArguments {
+    milestoneName: string;
+}
+
+export interface ObservableEvents<T> { // Normally T will be an EventEmitter
+    Events: T;
+}
+
+export interface StepStartedEventsEmitter {
+    on(event: 'stepStarted', listener: (args: StepStartedEventArguments) => void): this;
+    on(event: 'milestoneReached', listener: (args: MilestoneReachedEventArguments) => void): this;
+}
+
+export interface NavigatedToUserRequestedUrlEventsEmitter {
+    on(event: 'navigatedToUserRequestedUrl', listener: () => void): this;
+    once(event: 'navigatedToUserRequestedUrl', listener: () => void): this;
+}
+
+export class StepProgressEventsEmitter extends EventEmitter {
+    constructor(private readonly _nestedEmitters: [StepStartedEventsEmitter] = [] as [StepStartedEventsEmitter]) {
+        super();
+    }
+
+    public emitStepStarted(stepName: string): void {
+        this.emit(stepStartedEventName, { stepName: stepName } as StepStartedEventArguments);
+    }
+
+    public emitMilestoneReached(milestoneName: string): void {
+        this.emit(milestoneReachedEventName, { milestoneName: milestoneName } as MilestoneReachedEventArguments);
+    }
+
+    private subscribeToAllNestedEmitters(event: string, listener: Function): void {
+        for (const nestedEventEmitter of this._nestedEmitters) {
+            nestedEventEmitter.on(event as any, listener as any);
+        }
+    }
+
+    public on(event: string, listener: Function): this {
+        super.on(event, listener);
+        this.subscribeToAllNestedEmitters(event, listener);
+        return this;
+    }
+}
+
+/* Use to track the time executing each step during launch
```

Comment:
Is this usage comment necessary? It's incomplete since it doesn't say how to create Events or a reporter, and things like this quickly get out of date. Since this already includes code that uses these objects, I'm not sure this is needed.

Reply:
I removed the usage, and I left the sample telemetry

Reply:
That will get out of date too. Is it helpful?

Reply:
I think it's helpful to have a reference to what kind of telemetry we expect. If you don't like having this in the code, I can move this to a OneNote or something

Reply:
Removed